File: mac/tools/crosscheck/cuesheet.swift
import Foundation

struct T: CueTrack {
    var cueTitle: String
    var cueArtist: String = ""
    var cueKind: String = "Song"
    var cueSeconds: Double = 0.0
    var cueTicked: Bool = true
    var cueMissing: Bool = false
}

func pyList(_ items: [String]) -> String {
    "[" + items.map { "'\($0)'" }.joined(separator: ", ") + "]"
}
func pyOpt(_ i: Int?) -> String { i.map(String.init) ?? "None" }

var out: [String] = []
out.append("onair|\(CueSheet.onAir)")
out.append("missing|\(CueSheet.missingFile)")
out.append("end|\(CueSheet.endRow)")
out.append("empty|\(CueSheet.emptyRow)")

for value in [0, -1, 1, 59, 60, 61, 119, 120, 599, 3600, 3661] {
    out.append("length|\(value)|\(CueSheet.saidLength(Double(value)))")
}

let tracks: [CueTrack] = [
    T(cueTitle: "Opening Theme", cueArtist: "The Band", cueKind: "Song", cueSeconds: 185.0),
    T(cueTitle: "Station Ident", cueArtist: "", cueKind: "Drop", cueSeconds: 9.0),
    T(cueTitle: "Second Song", cueArtist: "Another Act", cueKind: "Song", cueSeconds: 212.4),
    T(cueTitle: "Not Ticked", cueArtist: "Nobody", cueKind: "Song", cueSeconds: 100.0, cueTicked: false),
    T(cueTitle: "Gone Missing", cueArtist: "A Ghost", cueKind: "Song", cueSeconds: 150.0, cueMissing: true),
    T(cueTitle: "Last One", cueArtist: "", cueKind: "Song", cueSeconds: 60.0),
]
let empty: [CueTrack] = []
let noneTicked: [CueTrack] = [
    T(cueTitle: "Only One", cueArtist: "X", cueKind: "Song", cueSeconds: 30.0, cueTicked: false)]
let allMissing: [CueTrack] = [
    T(cueTitle: "A", cueArtist: "", cueKind: "Song", cueSeconds: 10.0, cueMissing: true),
    T(cueTitle: "B", cueArtist: "", cueKind: "Song", cueSeconds: 20.0, cueMissing: true)]

let sets: [(String, [CueTrack])] = [
    ("full", tracks), ("empty", empty), ("none-ticked", noneTicked),
    ("all-missing", allMissing),
]

for (label, set) in sets {
    var playings: [Int?] = [nil]
    playings.append(contentsOf: (0..<set.count).map { Optional($0) })
    for playing in playings {
        for played in [0.0, 5.0, 9.99, 10.0, 10.01, 30.0] {
            let rows = CueSheet.build(tracks: set, playingIndex: playing, playedFor: played)
            let key = "\(label)|\(pyOpt(playing))|\(String(format: "%.2f", played))"
            out.append("rows|\(key)|\(rows.count)")
            for (i, row) in rows.enumerated() {
                out.append("  row|\(key)|\(i)|\(pyList(row.cells))")
                out.append("  idx|\(key)|\(i)|\(pyOpt(row.index))")
            }
            out.append("summary|\(key)|\(CueSheet.summary(tracks: set, rows: rows))")
            out.append("next|\(key)|\(CueSheet.nextFew(rows))")
            out.append("next1|\(key)|\(CueSheet.nextFew(rows, howMany: 1))")
            out.append("next9|\(key)|\(CueSheet.nextFew(rows, howMany: 9))")
        }
    }
}

for grace in [0.0, 1.0, 10.0, 60.0] {
    let rows = CueSheet.build(tracks: tracks, playingIndex: 1, playedFor: 5.0, grace: grace)
    out.append("grace|\(String(format: "%.1f", grace))|\(pyList(rows.map(\.title)))")
}

let shown = CueSheet.build(tracks: tracks, playingIndex: nil)
let wanted = CueSheet.build(tracks: tracks, playingIndex: 1, playedFor: 30.0)
var focuses: [String?] = [nil]
focuses.append(contentsOf: shown.map { Optional($0.title) })
focuses.append("Nothing By This Name")
focuses.append(CueSheet.endRow)
for focused in focuses {
    let got = CueSheet.applyChanges(shown: shown, wanted: wanted, focusedTitle: focused)
    out.append("apply|\(focused ?? "None")|\(pyList(got.map(\.title)))")
}

var got = CueSheet.applyChanges(shown: wanted, wanted: shown, focusedTitle: "Opening Theme")
out.append("apply-back|\(pyList(got.map(\.title)))")
got = CueSheet.applyChanges(shown: [], wanted: wanted, focusedTitle: "Opening Theme")
out.append("apply-fresh|\(pyList(got.map(\.title)))")
print(out.joined(separator: "\n"))

File: mac/tools/crosscheck/cuesheet.py
# The repository root is on the path already, put there by cross_check.py.
#
# What is coming up next. Which rows are in the list, in which order, what each
# one says and what the line under them says, plus the rule that decides which
# rows may leave while somebody is standing on one.
#
# That last rule is the whole of this case. It is what stops a screen reader
# stopping mid sentence and reading out a track the presenter did not choose,
# on air, at the moment a song changes, and it is pure list arithmetic on both
# platforms. It has to come out the same.
from dropdeck import cuesheet as F


class T:
    def __init__(self, title, artist="", kind="Song", duration=0.0,
                 ticked=True, missing=False):
        self.title = title
        self.artist = artist
        self.kind = kind
        self.duration = duration
        self.ticked = ticked
        self.missing = missing


out = []
out.append("onair|%s" % F.ON_AIR)
out.append("missing|%s" % F.MISSING)
out.append("end|%s" % F.END_ROW)
out.append("empty|%s" % F.EMPTY_ROW)

for value in (0, -1, 1, 59, 60, 61, 119, 120, 599, 3600, 3661):
    out.append("length|%d|%s" % (value, F.said_length(value)))

TRACKS = [
    T("Opening Theme", "The Band", "Song", 185.0),
    T("Station Ident", "", "Drop", 9.0),
    T("Second Song", "Another Act", "Song", 212.4),
    T("Not Ticked", "Nobody", "Song", 100.0, ticked=False),
    T("Gone Missing", "A Ghost", "Song", 150.0, missing=True),
    T("Last One", "", "Song", 60.0),
]
EMPTY = []
NONE_TICKED = [T("Only One", "X", "Song", 30.0, ticked=False)]
ALL_MISSING = [T("A", "", "Song", 10.0, missing=True),
               T("B", "", "Song", 20.0, missing=True)]

SETS = [("full", TRACKS), ("empty", EMPTY), ("none-ticked", NONE_TICKED),
        ("all-missing", ALL_MISSING)]

for label, tracks in SETS:
    for playing in [None] + list(range(len(tracks))):
        for played in (0.0, 5.0, 9.99, 10.0, 10.01, 30.0):
            rows = F.build(tracks, playing_index=playing, played_for=played)
            key = "%s|%s|%.2f" % (label, playing, played)
            out.append("rows|%s|%d" % (key, len(rows)))
            for i, row in enumerate(rows):
                out.append("  row|%s|%d|%s" % (key, i, list(row.cells())))
                out.append("  idx|%s|%d|%s" % (key, i, row.index))
            out.append("summary|%s|%s" % (key, F.summary(tracks, rows)))
            out.append("next|%s|%s" % (key, F.next_few(rows)))
            out.append("next1|%s|%s" % (key, F.next_few(rows, 1)))
            out.append("next9|%s|%s" % (key, F.next_few(rows, 9)))

# A grace that is not the default, so the number is really being read.
for grace in (0.0, 1.0, 10.0, 60.0):
    rows = F.build(TRACKS, playing_index=1, played_for=5.0, grace=grace)
    out.append("grace|%.1f|%s" % (grace, [r.title for r in rows]))

# And the rule about the row under the cursor.
shown = F.build(TRACKS, playing_index=None)
wanted = F.build(TRACKS, playing_index=1, played_for=30.0)
for focused in ([None] + [r.title for r in shown]
                + ["Nothing By This Name", F.END_ROW]):
    got = F.apply_changes(shown, wanted, focused)
    out.append("apply|%s|%s" % (focused, [r.title for r in got]))

# The same, the other way: the cursor is on a row that is only in `wanted`.
got = F.apply_changes(wanted, shown, "Opening Theme")
out.append("apply-back|%s" % [r.title for r in got])
# And on an empty displayed list, which is what the first refresh looks like.
got = F.apply_changes([], wanted, "Opening Theme")
out.append("apply-fresh|%s" % [r.title for r in got])
print("\n".join(out))
